### PR TITLE
Bump k8s apis to 1.27, refactor to support contexts

### PR DESCRIPTION
**Pull request by lsierant**
This PR contains multiple changes related to bumping kube and controller-gen APIs: 

* Bumped kube apis to v0.27.12
* Bumped controller-runtime v0.15.3 ([changes were huge and breaking](https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.15.0))

Due to numerous places where context was added in the API, I decided to go all in and refactor the whole codebase to properly support contexts. There are no dangling context.TODO or context.Background in random places anymore.

wait.Poll* functions were migrated away from the old, deprecated variants to the new wait.PollUntilContextTimeout

I've left few comments "api bump" in places where the diff is not a simple context change and was due to API changes.

### CODE REVIEW

File: controllers/replica_set_controller.go [hunk 1]
```diff
@@ -80,8 +78,8 @@ func (r *ReplicaSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		WithOptions(controller.Options{MaxConcurrentReconciles: 3}).
 		For(&mdbv1.MongoDBCommunity{}, builder.WithPredicates(predicates.OnlyOnSpecChange())).
-		Watches(&source.Kind{Type: &corev1.Secret{}}, r.secretWatcher).
```
Comment (lsierant) [inline controllers/replica_set_controller.go:83]: api bump

File: go.mod [hunk 2]
```diff
@@ -14,29 +14,29 @@ require (
 	go.mongodb.org/mongo-driver v1.14.0
 	go.uber.org/zap v1.27.0
 	gopkg.in/natefinch/lumberjack.v2 v2.2.1
-	k8s.io/api v0.26.10
-	k8s.io/apimachinery v0.26.10
-	k8s.io/client-go v0.26.10
-	sigs.k8s.io/controller-runtime v0.14.7
+	k8s.io/api v0.27.12
```
Comment (lsierant) [inline go.mod:17]: api bump

File: pkg/kube/client/mocked_client.go [hunk 3]
```diff
@@ -31,6 +32,14 @@ type mockedClient struct {
 	backingMap map[reflect.Type]map[k8sClient.ObjectKey]k8sClient.Object
 }
 
+func (m mockedClient) GroupVersionKindFor(obj runtime.Object) (schema.GroupVersionKind, error) {
```
Comment (lsierant) [inline pkg/kube/client/mocked_client.go:35]: api bump

File: pkg/kube/client/mocked_manager.go [hunk 4]
```diff
@@ -36,6 +36,10 @@ func NewManagerWithClient(c k8sClient.Client) *MockedManager {
 	return &MockedManager{Client: NewClient(c)}
 }
 
+func (m *MockedManager) GetHTTPClient() *http.Client {
```
Comment (lsierant) [inline pkg/kube/client/mocked_manager.go:39]: api bump

File: test/e2e/util/mongotester/mongotester.go [hunk 5]
```diff
@@ -244,16 +246,16 @@ func (m *Tester) connectivityCheck(shouldSucceed bool, opts ...OptionApplier) fu
...
 
-		if err := m.ensureClient(clientOpts...); err != nil {
+		if err := m.ensureClient(ctx, clientOpts...); err != nil {
 			t.Fatal(err)
 		}
 
 		attempts := 0
 		// There can be a short time before the user can auth as the user
-		err := wait.Poll(connectivityOpts.IntervalTime, connectivityOpts.TimeoutTime, func() (done bool, err error) {
+		err := wait.PollUntilContextTimeout(ctx, connectivityOpts.IntervalTime, connectivityOpts.TimeoutTime, false, func(ctx context.Context) (done bool, err error) {
```
Comment (lsierant) [inline test/e2e/util/mongotester/mongotester.go:258]: api bump

**Review by slaskawi — APPROVED** [review]

**Comment by slaskawi** [comment]
LGTM assuming the CI is green

**Review by mircea-cosbuc — APPROVED** [review]
LGTM

**Pull request merged**